File: incidents/src/main/java/fr/openent/incidents/service/PartnerService.java
package fr.openent.incidents.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public interface PartnerService {
    /**
     * get partners
     *
     * @param structureId   Structure Identifier
     * @param handler       Function handler returning data
     */
    void get(String structureId, Handler<Either<String, JsonArray>> handler);

    /**
     * create partner
     *
     * @param partnerBody  partnerBody fetched
     * @param handler      Function handler returning data
     */
    void create(JsonObject partnerBody, Handler<Either<String, JsonObject>> handler);

    /**
     * update partner
     *
     * @param partnerBody  partnerBody fetched
     * @param handler      Function handler returning data
     */
    void put(JsonObject partnerBody, Handler<Either<String, JsonObject>> handler);

    /**
     * delete partner
     *
     * @param partnerId partner Identifier
     * @param handler   Function handler returning data
     */
    void delete(Integer partnerId, Handler<Either<String, JsonObject>> handler);
}

File: incidents/src/main/resources/public/ts/services/PartnerService.ts
import {ng} from 'entcore'
import { http, HttpResponse } from 'entcore-toolkit';

export interface Partner {
    id: number;
    structure_id?: string;
    label: string;
    hidden: boolean;
    used?: boolean;
}

export interface PartnerRequest {
    id?: number;
    structureId?: string,
    label?: string;
    hidden?: boolean;
}


export interface PartnerService {
    get(structureId: string): Promise<Partner[]>;
    create(partnerBody: PartnerRequest): Promise<HttpResponse>;
    update(partnerBody: PartnerRequest): Promise<HttpResponse>;
    delete(partnerId: number): Promise<HttpResponse>;
}

export const partnerService : PartnerService = {
    get: async (structureId: string): Promise<Partner[]> => {
        try {
            const {data} = await http.get(`/incidents/partners?structureId=${structureId}`);
            return data;
        } catch (err) {
            throw err;
        }
    },

    create: async (partnerBody: PartnerRequest): Promise<HttpResponse> => {
        return await http.post(`/incidents/partner`, partnerBody);
    },

    update: async (partnerBody: PartnerRequest): Promise<HttpResponse> => {
        return await http.put(`/incidents/partner`, partnerBody);
    },

    delete: async (partnerId: number): Promise<HttpResponse> => {
        return await http.delete(`/incidents/partner?id=${partnerId}`);
    },
};

export const PartnerService = ng.service('PartnerService', (): PartnerService => partnerService);
